src/coreclr/src/gc/unix/cgroup.cpp

```diff
@@ -59,96 +67,116 @@ class CGroup
+            return GetCGroup1CpuLimit(val);
+        else if (s_cgroup_version == 2)
+            return GetCGroup2CpuLimit(val);
+        else
+        {
+            assert(!"Unknown cgroup version.");
+            return false;
+        }
+    }
 
-        size_t len = strlen(s_memory_cgroup_path);
-        len += strlen(MEM_USAGE_FILENAME);
-        mem_usage_filename = (char*)malloc(len+1);
-        if (mem_usage_filename == nullptr)
-            return result;
+private:
+    static int FindCGroupVersion()
+    {
+        int version = 0;
+        char *line = nullptr;
+        size_t lineLen = 0;
+        char *filesystemType = nullptr;
 
-        strcpy(mem_usage_filename, s_memory_cgroup_path);
-        strcat(mem_usage_filename, MEM_USAGE_FILENAME);
-        result = ReadMemoryValueFromFile(mem_usage_filename, &temp);
-        if (result)
+        FILE *mountinfofile = fopen(PROC_MOUNTINFO_FILENAME, "r");
+        if (mountinfofile == nullptr)
+            goto done;
+
+        while (getline(&line, &lineLen, mountinfofile) != -1)
         {
-            if (temp > std::numeric_limits<size_t>::max())
+            // See man page of proc to get format for /proc/self/mountinfo file
+            char *separatorChar = strstr(line, " - ");
+            if (separatorChar == nullptr)
             {
-                *val = std::numeric_limits<size_t>::max();
+                assert(!"Failed to parse mount info file contents.");
+                goto done;
             }
-            else
+            char *filesystemTypeStart = separatorChar + strlen(" - ");
+            if (filesystemTypeStart == nullptr)
             {
-                *val = (size_t)temp;
+                assert(!"Failed to parse mount info file contents.");
+                goto done;
             }
-        }
-        free(mem_usage_filename);
-        return result;
-    }
 
-    static bool GetCpuLimit(uint32_t *val)
-    {
-        long long quota;
-        long long period;
-        double cpu_count;
+            char *filesystemTypeEnd = strstr(filesystemTypeStart, " ");
+            if (filesystemTypeEnd == nullptr)
+            {
+                assert(!"Failed to parse mount info file contents.");
+                goto done;
+            }
 
-        quota = ReadCpuCGroupValue(CFS_QUOTA_FILENAME);
-        if (quota <= 0)
-            return false;
+            size_t filesystemTypeLen = filesystemTypeEnd - filesystemTypeStart;
+            filesystemType = filesystemTypeStart;
 
-        period = ReadCpuCGroupValue(CFS_PERIOD_FILENAME);
-        if (period <= 0)
-            return false;
-
-        // Cannot have less than 1 CPU
-        if (quota <= period)
-        {
-            *val = 1;
-            return true;
+            if (strncmp(filesystemType, "cgroup", filesystemTypeLen) == 0)
+            {
+                version = 1;
+                goto done;
+            }
+            else if (strncmp(filesystemType, "cgroup2", filesystemTypeLen) == 0)
```

Comment:
Although `filesystemTypeLen`is calculated from the string you just read from the file, I'd test for `cgroup2` first, as it's the longest string.

Reply:
Fixed.